codegen: implement `For` ranges

amscanne opened the pull request:
codegen: implement `For` ranges

This adds the `..` syntax for `For` loops, allows two expressions (as
long as they are unsigned types).

Signed-off-by: Adin Scannell <amscanne@meta.com>

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2563,30 +2570,41 @@ void SemanticAnalyser::visit(For &f)
+  } else if (auto *range = f.iterable.as<Range>()) {
+    if (is_final_pass()) {
+      if (!range->start.type().IsIntTy()) {
+        range->addError() << "Loop range requires integer start";
```
nit
```suggestion
        range->addError() << "Loop range requires an integer for the start";
```
(same for the end below)

jordalgo commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -4687,22 +4699,136 @@ llvm::Function *CodegenLLVM::createForEachMapCallback(For &f, llvm::Type *ctx_t)
 
+ScopedExpr CodegenLLVM::visit(For &f, Range &range)
+{
+  // Evaluate our starting an endpoint values.
```
```suggestion
  // Evaluate our starting and endpoint values.
```

amscanne replied:
Done.

amscanne commented on src/bpffeature.h:
```diff
@@ -124,6 +124,7 @@ class BPFfeature {
   DEFINE_HELPER_TEST(for_each_map_elem, libbpf::BPF_PROG_TYPE_KPROBE);
   DEFINE_HELPER_TEST(get_ns_current_pid_tgid, libbpf::BPF_PROG_TYPE_KPROBE);
   DEFINE_HELPER_TEST(map_lookup_percpu_elem, libbpf::BPF_PROG_TYPE_KPROBE);
+  DEFINE_HELPER_TEST(loop, libbpf::BPF_PROG_TYPE_KPROBE);
```
Sure, will do.

amscanne commented on tests/semantic_analyser.cpp:
```diff
@@ -4535,6 +4535,118 @@ BEGIN { @map[0] = count(); for ($kv : @map) { print($kv); } }
+)");
+}
+
+TEST(semantic_analyser, for_range_control_flow)
```
Sure, I'll add it to the stack here. It does seem pretty trivial to support break/continue. Supporting `return` is a bit more complex, so it'll be another separate change (maybe post-poned a bit more).

amscanne replied:
This will be separate.

amscanne commented on tests/parser.cpp:
```diff
@@ -2931,6 +2931,57 @@ BEGIN { for (@kv : @map) { } }
 )");
 }
 
+TEST(Parser, for_range)
```
Let me know. I'll have more tests with the control flow stuff coming shortly.

jordalgo commented:
Also, probably good to call out in the docs that even though `$i` is being mutated the range is a snapshot and will not be updated during iteration e.g. this prints 0 through 9 instead of 0 through 4.
```
i:s:1 {
  $a = 10;
   for ($i : 0..$a) {
     print($i); 
     $a--;
    }
}
```

amscanne commented:
> @amscanne FYI: I ran into a SEGV on my ASAN build for this PR when running this program:
> 
> ```
> BEGIN { @a = 1; }
> i:s:1 { 
>   for ($i : 0..@a) {
>     print($i);
>     @a += 1;
>   }
> }
> ```
> 
> Can you repro?
> 
> Ok found the issue. You need to update the map_sugar visit. Here is the updated code:
> 
> ```
> void MapDefaultKey::visit(For &for_loop)
> {
>   if (auto *map = for_loop.iterable.as<Map>()) {
>     if (!check(*map, true)) {
>       map->addError() << map->ident
>                       << " has no explicit keys (scalar map), and "
>                          "cannot be used for iteration";
>     }
>   } else if (auto *range = for_loop.iterable.as<Range>()) {
>     visit(range->start);
>     visit(range->end);
>   }
>   visit(for_loop.stmts);
> }
> ```
> 
> Probably good to add a runtime test for this too.

I think I saw this just before you edited. I also found it and fixed it, and added a variant of your test. Good catch.

jordalgo reviewed (approved):
Had a final docs comment but non-blocking. Nice work!

amscanne commented:
I've updated the docs with information about mutability of the bounds and an example.
